project/src/main/world/creature/CreatureVisuals.tscn

```diff
@@ -252,14 +328,90 @@ shader_param/green = Color( 0, 0, 0, 1 )
 shader_param/blue = Color( 0, 0, 0, 1 )
 shader_param/black = Color( 0, 0, 0, 1 )
 
-[sub_resource type="ShaderMaterial" id=24]
+[sub_resource type="Shader" id=123]
+code = "/*
+ * A shader which takes an RGB image, and replaces the red component with one color, the green component with a second
+ * color, and the blue component with a third color. Black pixels are also replaced with a fourth color. A pixel
+ * which is mostly red but a little green will be replaced with the first and second colors.
+ *
+ * This approach allow a sprite to be recolored with up to four channels, and the recoloring can be handled by the
+ * graphics card. It's performant and looks reasonable even if the sprite is resized before being passed off to the
+ * graphics card.
+ */
+shader_type canvas_item;
+
+// the replacement colors
+uniform vec4 red : hint_color;
+uniform vec4 green : hint_color;
+uniform vec4 blue : hint_color;
+uniform vec4 black : hint_color;
+
+void fragment() {
+	vec4 rgba_in = texture(TEXTURE, UV);
+	
+	// color defaults to the black replacement color
+	vec4 rgb_out = black;
+	
+	// mix in other colors based on the red, green and blue components of the source image
+	rgb_out = mix(rgb_out, red.rgba, rgba_in.r);
+	rgb_out = mix(rgb_out, green.rgba, rgba_in.g);
+	rgb_out = mix(rgb_out, blue.rgba, rgba_in.b);
+	
+	// Assign final color for the pixel, and preserve transparency
+	COLOR = vec4(rgb_out.rgb, rgba_in.a * rgb_out.a);
+}"
+custom_defines = ""
+
+[sub_resource type="ShaderMaterial" id=124]
 resource_local_to_scene = true
-shader = ExtResource( 1 )
+shader = SubResource( 123 )
 shader_param/red = Color( 0, 0, 0, 1 )
 shader_param/green = Color( 0.952941, 0.572549, 0.454902, 1 )
 shader_param/blue = null
 shader_param/black = Color( 0, 0, 0, 1 )
 
+[sub_resource type="Shader" id=117]
+code = "/*
```

Comment:
Do we need this same shader code repeated ~30 times? I know we need the parameters to be different but does the shader need to be different too?

Reply:
The shader doesn't need to be different, just the material. I've removed this shader code.